helion/autotuner/ucb_pattern_search.py

```diff
@@ -0,0 +1,355 @@
+from __future__ import annotations
+
+import math
+import random
+from typing import TYPE_CHECKING
+
+import torch
+
+from .. import exc
+from .base_search import FlatConfig
+from .base_search import PopulationMember
+from .base_search import performance
+from .config_fragment import PowerOfTwoFragment
+from .effort_profile import PATTERN_SEARCH_DEFAULTS
+from .fragment_encoder import ConfigEncoder
+from .pattern_search import PatternSearch
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+    from collections.abc import Sequence
+
+    from ..runtime.config import Config
+    from ..runtime.kernel import BoundKernel
+
+
+import operator
+
+from botorch.acquisition import UpperConfidenceBound
+from botorch.fit import fit_gpytorch_mll
+from botorch.models import MixedSingleTaskGP
+from gpytorch.mlls import ExactMarginalLogLikelihood
```

Comment:
Optional deps need to be imported in local scope.

How stable/widely used is this dep?